Securitypolicyviolationevent sidebar (#35384)
mdn/content: files/en-us/web/api/document/securitypolicyviolation_event/index.md

The **`securitypolicyviolation`** event is fired when a [Content Security Policy](/en-US/docs/Web/HTTP/CSP) is violated.

The event is fired on the document when there is a violation of the document CSP policy (and may also bubble from elements in the document).

This event [bubbles](/en-US/docs/Learn/JavaScript/Building_blocks/Event_bubbling) to the {{domxref("Window")}} object, and is [composed](/en-US/docs/Web/API/Event/composed).

> [!NOTE]
> You should add the handler for this event to a top level object (i.e. {{domxref("Window")}} or {{domxref("Document")}}).

> > While the property exists in HTML elements, you can't assign a handler to the property until the elements have been loaded, by which time this event will already have fired.

Reviewer: I'm not sure this is true for dynamically constructed elements, such as `document.createElement("img")`. Not what this PR changed though.

Author: So this is saying if your static HTML element would trigger a violation, perhaps because it loads a restricted img, by the time you get to add your handler the event will have already fired? Makes sense, though I guess if you declare the handler inline that might be assigned and ready before the violation triggers. That could be browser dependent.

You're saying that you can add a dynamically constructed element with a handler, but the the policy violation isn't triggered until here is a global context - when you add it to the document. So that should work.

Would this fix it "enough"? I don't want to mess with this too much.

```suggestion
> This is because you can't define a handler for elements in your HTML file until they are fully loaded, by which time the event will already have fired.
```
